clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -577,33 +586,42 @@ private ListOffsetResult fetchOffsetsByTimes(Map<TopicPartition, Long> timestamp
      */
     public Map<TopicPartition, List<ConsumerRecord<K, V>>> fetchedRecords() {
         Map<TopicPartition, List<ConsumerRecord<K, V>>> fetched = new HashMap<>();
+        Queue<CompletedFetch> pausedCompletedFetches = new ArrayDeque<>();
```

Comment:
I have a few suggestions to simplify this a little bit:

1. The `PartitionRecords` parses records in a streaming fashion. The constructor doesn't do any parsing, so there's no harm constructing it immediately when we find a `CompletedFetch` for a paused partition. So rather than leaving the completed fetch inside `completedFetches`, it seems like we could keep them all in the other queue (maybe `pausedCompletedFetches` instead of `parsedFetchesCache`)? As a matter of fact, I think we can change the type of `completedFetches` to `ConcurrentLinkedQueue<PartitionRecords>`.

2. Below we have logic to basically iterate through the `parsedFetchesCache` and move all the partitions which are still paused to the end. Would it be simpler to leave the collection in the same order and 

3. Once a partition is unpaused, does it make sense to just move it back to `completedFetches`? Then 

So then the logic would be just to iterate through the paused data and check for unpaused partitions which can be moved over to `completedFetches`. Then everything is done as usual.

Reply:
Thanks for the review @hachikuji !  I think it would be nice to have the single `completedFetches` of type `ConcurrentLinkedQueue<PartitionRecords>`.  To do this we would have to change the fetch parsing workflow a little bit, but I'm not sure what ramifications this may have.  I've made some assumptions below I hope you can confirm before I refactor the PR.

* In the callback for `sendFetches` we would immediately call `parseCompletedFetch` to create a `PartitionRecords` and add it to the `completedFetches`, instead of calling `parseCompletedFetch` later in `fetchedRecords`.  The javadoc comment on `parseCompletedFetch` method suggests that at one point in the past this may have been the case: "The callback for fetch completion".  If we move the call here I don't think there's any reason to have a `CompletedFetch` type at all.
* The callsite to `parseCompletedFetch` is wrapped in exception handling which discards the fetch in some scenarios when an exception is raised.  If we move the callsite to `sendFetches` then I assume we would move the exception handling along with it.
* In `parseCompletedFetch` if the partition is unfetchable then no records are returned.  Based on comments in this method this could be caused by a rebalance that happened while a fetch request was in flight, or when a partition is paused.  We would need to change this so that the `PartitionRecords` are still created when a partition is paused, but not in other unfetchable scenarios.

Reply:
I tried refactoring the PR with the assumptions I made above, but it began to reveal a lot of regression bugs when running `FetcherTest`.  I think my idea about moving the `parseCompletedFetch` callsite to `sendFetches` seems to miss a lot of nuance accounted for in the tests.

Reply:
I created a PR on my fork with the refactor: https://github.com/seglo/kafka/pull/3/files

Reply:
@seglo Yeah, this is a little trickier than I suggested. The draft looks like it's heading in the right direction. One thought I had is to take the `parseCompletedFetch` logic and turn it into an `initializePartitionRecords` function. So when we first construct `PartitionRecords`, we consider it uninitialized. Then the logic in `fetchedRecords` could look something like this:

```java
nextInLineRecords = completedFetches.peek();
if (nextInLineRecords.notInitialized()) {
  try {
    initializePartitionRecords(nextInLineRecords);
  } catch (Exception e) {
  }
}
// now check in nextInLineRecords has any data to fetch
```
The benefit is that the same logic is still executed from the same location. Would that help?

Reply:
@hachikuji Yes, that suggestion helped, thanks.  I no longer have any `FetcherTest` regression failures.  I split up `parseCompletedFetch` into:

* A `parseCompletedFetch` that always returns a `PartitionRecords`, but does no other validation.  Called from `sendFetches`.
* An `initializePartitionRecords` which does all the validation that `parseCompletedFetch` did before.  Called from `fetchedRecords` where exceptions are handled.

`completedFetches` is now used not just as a queue, but as a cache.  The queue usage semantics aren't really applicable anymore (we no longer `peek` or `poll`), but since we want the ordering I don't think there's a better datastructure in `java.util.concurrent` to use.  

I've updated this PR.

Reply:
I made an update to cache each paused completed fetch for the lifetime of a call to `fetchRecords` and then add them back at the end so that `completedFetches.poll()` can be used instead of removing by object reference `completedFetches.remove(records)`.  I'm not sure if this is any more efficient, but it preserves the original implementation of `fetchedRecords` better.  The change is in this commit: https://github.com/apache/kafka/pull/6988/commits/7cb79431993db8bd867311dbab50f80e60d68e6a .